p2p/src/net/types/mod.rs

```diff
@@ -211,3 +211,7 @@ pub enum ValidationResult {
     /// Message is not invalid but it shouldn't be forwarded to other peers
     Ignore,
 }
+
+pub trait GetIp {
```

Comment:
Why don't you define an address type here as an associated type? Wouldn't that solve the problem you're having with channels?

Reply:
I'm not sure. Yes, some kind of an address can be introduced specifically for channels, but it would still have the same issues: it needs to be randomly generated (this can be done for `IpAddr` too) and it needs to be "persistent".